internal/gps/source_cache_bolt.go

```diff
@@ -0,0 +1,437 @@
+//
+//	Sub-Bucket: "versions:<timestamp>"
+//	Keys: "<sequence_number>"
+//	Values: "branch:<branch>", "defaultBranch:<branch>", "ver:<version>"
+type singleSourceCacheBolt struct {
+	ProjectRoot
+	db     *bolt.DB
+	epoch  int64       // getters will not return values older than this unix timestamp
+	logger *log.Logger // info logging
+}
+
+// newBoltCache returns a new singleSourceCacheBolt backed by a project's BoltDB file under the cache directory.
+func newBoltCache(cd string, pi ProjectIdentifier, epoch int64, logger *log.Logger) (*singleSourceCacheBolt, error) {
+	path := sourceCachePath(cd, pi.normalizedSource()) + ".db"
+	dir := filepath.Dir(path)
+	if fi, err := os.Stat(dir); os.IsNotExist(err) {
+		if err := os.MkdirAll(dir, os.ModeDir|os.ModePerm); err != nil {
+			return nil, errors.Wrapf(err, "failed to create source cache directory: %s", dir)
+		}
+	} else if err != nil {
+		return nil, errors.Wrapf(err, "failed to check source cache directory: ", dir)
+	} else if !fi.IsDir() {
+		return nil, errors.Wrapf(err, "source cache path is not directory: %s", dir)
+	}
+	db, err := bolt.Open(path, 0600, &bolt.Options{Timeout: 1 * time.Second})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to open BoltDB cache file %q", path)
+	}
+	return &singleSourceCacheBolt{
+		ProjectRoot: pi.ProjectRoot,
+		db:          db,
+		epoch:       epoch,
+		logger:      logger,
+	}, nil
+}
+
+// close releases all database resources.
+// Must not be called concurrently with any other methods.
+func (s *singleSourceCacheBolt) close() error {
+	return errors.Wrapf(s.db.Close(), "error closing Bolt database %q", s.db.String())
+}
+
+func (s *singleSourceCacheBolt) setManifestAndLock(rev Revision, ai ProjectAnalyzerInfo, m Manifest, l Lock) {
+	err := s.updateBucket("rev:"+string(rev), func(b *bolt.Bucket) error {
+		pre := "info:" + ai.String() + ":"
+		if err := cachePrefixDelete(b, pre); err != nil {
+			return err
+		}
+		info, err := b.CreateBucket(cacheTimestampedKey(pre, time.Now()))
+		if err != nil {
+			return err
+		}
+
+		// Manifest
+		mb, err := info.CreateBucket([]byte("manifest"))
+		if err != nil {
+			return err
+		}
+		if err := cachePutManifest(mb, m); err != nil {
+			return errors.Wrap(err, "failed to put manifest")
+		}
+		if l == nil {
+			return nil
+		}
+
+		// Lock
+		lb, err := info.CreateBucket([]byte("lock"))
+		if err != nil {
+			return err
+		}
+		return errors.Wrap(cachePutLock(lb, l), "failed to put lock")
+	})
+	if err != nil {
+		s.logger.Println(errors.Wrapf(err, "failed to cache manifest/lock for revision %q, analyzer: %v", rev, ai))
+	}
+}
+
+func (s *singleSourceCacheBolt) getManifestAndLock(rev Revision, ai ProjectAnalyzerInfo) (m Manifest, l Lock, ok bool) {
+	err := s.viewBucket("rev:"+string(rev), func(b *bolt.Bucket) error {
+		info := cacheFindLatestValid(b, "info:"+ai.String()+":", s.epoch)
+		if info == nil {
+			return nil
+		}
+
+		// Manifest
+		mb := info.Bucket([]byte("manifest"))
```

Comment:
i'm not sure if the go compiler optimizes out the string literal conversion here, but if we can't be sure, let's use package-level vars to avoid realloc, etc.

Reply:
I sure hope it does, although now I wonder if it would have to prove it's not mutated by the method in order to share a single reference. Regardless, this will likely go away anyways, between dropping the time-stamp and using an encoding like protobuf.